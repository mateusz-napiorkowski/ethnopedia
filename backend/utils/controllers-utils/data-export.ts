import Artwork from "../../models/artwork";

export const fillRow: any = (keys: any, categories: any) => {
    const rowdata: any = {}
    keys.forEach((key: any) => {
        rowdata[key] = findValue(categories, key.split("."))
    });
    return rowdata
}

const findValue: any = (subcategories: any, keySplit: any) => {
    let v: string | undefined = undefined
    const categoryDepth = keySplit.length
    if(categoryDepth > 1) {
        const categoryPrefix = keySplit[0]
        for(const subcategory of subcategories) {
            if(subcategory.name == categoryPrefix) {
                v = findValue(subcategory.subcategories, keySplit.slice(1))
                if(v !== undefined) return v
            }
        }
    } else if (categoryDepth == 1) {
        if(subcategories !== undefined) {
            for(const subcategory of subcategories) {
                if(subcategory.name == keySplit[0]) {
                    const v_temp: Array<string> = []
                    for(const subcategoryValue of subcategory.values) {
                        v_temp.push(subcategoryValue)
                    }
                    return v_temp.join(";")
                }
            }
        }  
    }
    return ""
}

const getNestedCategories = ((prefix: string, subcategories: any) => {
    const nestedCategories: Array<string> = []
    if(subcategories !== undefined) {
        for(const subcategory of subcategories){
            nestedCategories.push(`${prefix}${subcategory.name}`)
            nestedCategories.push(...getNestedCategories(`${prefix}${subcategory.name}.`, subcategory.subcategories))
        }
    }  
    return nestedCategories
})

export const getAllCategories = async (collectionName: any) => {
    const records = await Artwork.find({ collectionName: collectionName })
    const allCategories: Array<string> = []
    records.forEach((record:any) => {
        for(const category of record.categories){
            allCategories.push(category.name)
            allCategories.push(...getNestedCategories(`${category.name}.`, category.subcategories))
        }
    });
    const allCategoriesUnique = allCategories.filter((value, index, array) => {
        return array.indexOf(value) === index;
    })
    return allCategoriesUnique
}