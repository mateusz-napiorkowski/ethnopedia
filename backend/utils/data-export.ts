import { subcategoryData } from "./interfaces"

const findValue: any = (subcategories: Array<subcategoryData>, categoryNameSplitByDot: Array<string>) => {
    const categoryDepth = categoryNameSplitByDot.length
    const [topmostParentCategoryName, ...categoryNameWithoutTopmostPart] = categoryNameSplitByDot;
    if(categoryDepth > 1) {
        const matchingSubcategory = subcategories.find(subcategory => subcategory.name === topmostParentCategoryName);
        return matchingSubcategory ? findValue(matchingSubcategory.subcategories, categoryNameWithoutTopmostPart) : '';
    }
    const matchingCategory = subcategories.find(category => category.name === topmostParentCategoryName);
    return matchingCategory ? matchingCategory.values.join(';') : '';
}

export const fillRow = (keys: Array<string>, categories: Array<subcategoryData>) => {
    const rowdata: any = {}
    keys.forEach(key => {
        rowdata[key] = findValue(categories, key.split("."))
    });
    return rowdata
}