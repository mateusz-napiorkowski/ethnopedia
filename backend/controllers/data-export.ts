import { Request, Response } from "express"
import excelJS from "exceljs"
import Artwork from "../models/artwork";
import { fillRow } from "../utils/data-export"
import { getAllCategories } from "../utils/categories";
import { constructAdvSearchFilter, constructQuickSearchFilter } from "../utils/artworks";
import CollectionCollection from "../models/collection";

export const getXlsxWithCollectionData = async (req: Request, res: Response) => {
    try {
        const collectionId = req.params.collectionId
        const collection = await CollectionCollection.findOne({_id: collectionId}).exec()
        if (collection == null)
            throw new Error(`Collection not found`)
        const collectionName = collection?.name as string

        const workbook = new excelJS.Workbook()
        const sheet = workbook.addWorksheet(collectionName)

        const columnNames = await getAllCategories([collectionId])
        sheet.columns = columnNames.map((name :string) => {return {header: name, key: name}})
        
        const records = await Artwork.find({collectionName: collectionName}).exec()
        records.forEach((record: any) => sheet.addRow(fillRow(columnNames, record.categories)))

        // //cell formatting
        sheet.columns.forEach((column)=>  {
            let maxLength = 0;
            column["eachCell"]!({ includeEmpty: true }, function (cell) {
                const columnLength = cell.value ? cell.value.toString().length : 10;
                if (columnLength > maxLength ) {
                    maxLength = columnLength;
                }
            });
            column.width = maxLength < 10 ? 10 : maxLength;
        });
        
        res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
        res.setHeader("Content-Disposition", "attachment");

        await workbook.xlsx.write(res)
        
        res.status(200).end()
    } catch (error) {
        const err = error as Error
        console.error(error)
        if (err.message === `Collection not found`)
            res.status(404).json({ error: err.message })
        else
            res.status(503).json({ error: `Database unavailable` })
    }
}

export const getXlsxWithArtworksData = async (req: Request, res: Response) => {
    try {
        const collectionIds = req.query.collectionIds

        const columnNames = typeof req.query.columnNames === "string" ? Array(req.query.columnNames) : req.query.columnNames as Array<string>
        
        const exportExtent = req.query.exportExtent
        if(!columnNames || !exportExtent || !collectionIds)
            throw new Error("Request is missing query params")

        const collections = await CollectionCollection.find({_id: {$in: collectionIds}}).exec()

        if (collections.length === 0)
            throw new Error(`Collection not found`)
        const collectionNames = collections.map(collection => collection.name as string);

        const workbook = new excelJS.Workbook()
        const worksheetName = collectionNames.length === 1 ? collectionNames[0] : "Arkusz1"
        const sheet = workbook.addWorksheet(worksheetName)
        sheet.columns = columnNames.map((name: string) => {return {header: name, key: name}})

        const selectedArtworks = req.query.selectedArtworks
        if(exportExtent === "selected" && !selectedArtworks) {    
            throw new Error("Request is missing query params")
        } else if(exportExtent === "selected" && selectedArtworks) {
            const records = await Artwork.find({collectionName: {$in: collectionNames}, _id: { $in: selectedArtworks}}).exec()
            records.forEach((record: any) => sheet.addRow(fillRow(columnNames, record.categories)))
        } else if(exportExtent === "searchResult") {
            const searchText = req.query.searchText
            const queryFilter = searchText ? await constructQuickSearchFilter(searchText, collectionIds as Array<string>, collectionNames) :
                await constructAdvSearchFilter(req.query, collectionNames, false)
            const records = await Artwork.find(queryFilter).exec()
            records.forEach((record: any) => sheet.addRow(fillRow(columnNames, record.categories)))
        } else {
            const records = await Artwork.find({collectionName: {$in: collectionNames}}).exec()
            records.forEach((record: any) => sheet.addRow(fillRow(columnNames, record.categories)))
        }
        
        //cell formatting
        sheet.columns.forEach((column) => {
            let maxLength = 0;
            column["eachCell"]!({ includeEmpty: true }, function (cell) {
                const columnLength = cell.value ? cell.value.toString().length : 10;
                if (columnLength > maxLength ) {
                    maxLength = columnLength;
                }
            });
            column.width = maxLength < 10 ? 10 : maxLength;
        });
        
        res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
        res.setHeader("Content-Disposition", "attachment");

        await workbook.xlsx.write(res)

        res.status(200).end()
    } catch (error) {
        const err = error as Error
        console.error(error)
        if (err.message === "Request is missing query params")
            res.status(400).json({ error: err.message })
        else if (err.message === `Collection not found`)
            res.status(404).json({ error: err.message })
        else
            res.status(503).json({ error: `Database unavailable` })
    }
}