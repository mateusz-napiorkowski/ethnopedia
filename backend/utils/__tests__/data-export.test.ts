import {describe, expect, test, jest, beforeEach} from "@jest/globals"
import { fillRow } from "../data-export";

describe('data-export util functions tests', () => {
    beforeEach(() => {
        jest.resetAllMocks()
    })

    test.each([
        {
            testName: 'fillRow test - functions returns correct data',
            keys: [
                'Tytuł',
                'Tytuł.Podtytuł',
                'Tytuł.Podtytuł.Podpodtytuł',
                'Artyści',
                'Rok',
                'Rok.Miesiąc',
                'Rok.Miesiąc.Dzień',
                'Tytuł.Znaczenie tytułu',
                'Rok.Kwartał',
                'Rok.Kwartał.Miesiąc',
                'Rok.Kwartał.Miesiąc.Dzień'
              ],
            categories: [
                {
                  name: 'Tytuł',
                  values: [ 'testowy' ],
                  subcategories: [
                    {
                      name: 'Podtytuł',
                      values: [ 'podtytuł testowy' ],
                      subcategories: [ { name: 'Podpodtytuł', values: [ 'podpodtytuł testowy' ], subcategories: [] } ]
                    }
                  ]
                },
                { name: 'Artyści', values: [ 'testowi' ], subcategories: [] },
                {
                  name: 'Rok',
                  values: [ '966' ],
                  subcategories: [
                    {
                      name: 'Miesiąc',
                      values: [ '12' ],
                      subcategories: [ { name: 'Dzień', values: [ '13' ], subcategories: [] } ]
                    }
                  ]
                }
            ]
        },
    ])(`$testName`,
        async ({keys, categories}) => {
            expect(fillRow(keys, categories)).toMatchSnapshot()
        }
    )
})