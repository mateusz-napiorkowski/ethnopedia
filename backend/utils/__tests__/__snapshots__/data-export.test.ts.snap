// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data-export util functions tests fillRow test - functions returns correct data 1`] = `
{
  "Artyści": "testowi",
  "Rok": "966",
  "Rok.Kwartał": "",
  "Rok.Kwartał.Miesiąc": "",
  "Rok.Kwartał.Miesiąc.Dzień": "",
  "Rok.Miesiąc": "12",
  "Rok.Miesiąc.Dzień": "13",
  "Tytuł": "testowy",
  "Tytuł.Podtytuł": "podtytuł testowy",
  "Tytuł.Podtytuł.Podpodtytuł": "podpodtytuł testowy",
  "Tytuł.Znaczenie tytułu": "",
}
`;
